Source/WebCore/html/HTMLMediaElement.cpp

```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
+                m_mediaSource = nullptr;
+                return false;
+            }
+            return true;
         }
-    }
 #endif
 #if ENABLE(MEDIA_STREAM)
-    if (!loadAttempted && m_mediaStreamSrcObject && !m_remotePlaybackConfiguration) {
-        loadAttempted = true;
-        ALWAYS_LOG(LOGIDENTIFIER, "loading media stream blob ", m_mediaStreamSrcObject->logIdentifier());
-        if (!player->load(m_mediaStreamSrcObject->protectedPrivateStream()))
-            mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
-    }
+        if (m_mediaStreamSrcObject && !m_remotePlaybackConfiguration) {
+            ALWAYS_LOG(LOGIDENTIFIER, "loading media stream blob ", m_mediaStreamSrcObject->logIdentifier());
+            return player->load(m_mediaStreamSrcObject->protectedPrivateStream());
+        }
 #endif
 
-    if (!loadAttempted && m_blob && !m_remotePlaybackConfiguration) {
-        loadAttempted = true;
-        ALWAYS_LOG(LOGIDENTIFIER, "loading generic blob");
-        if (!m_blobURLForReading.isEmpty())
-            ThreadableBlobRegistry::unregisterBlobURL(m_blobURLForReading);
-        m_blobURLForReading = { BlobURL::createPublicURL(document().protectedSecurityOrigin().ptr()), document().topOrigin().data() };
-        ThreadableBlobRegistry::registerBlobURL(document().protectedSecurityOrigin().ptr(), document().policyContainer(), m_blobURLForReading, m_blob->url());
+        return player->load(url, contentType, keySystem, !!m_remotePlaybackConfiguration);
+    };
 
-        if (!player->load(m_blobURLForReading, contentType.isEmpty() ? ContentType { m_blob->type() } : contentType, keySystem, !!m_remotePlaybackConfiguration))
-            mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
-    }
+    if (contentType.isEmpty() && url.protocolIsData())
+        contentType = ContentType(mimeTypeFromDataURL(url.string()));
 
-    if (!loadAttempted && !player->load(url, url.protocolIsBlob() && contentType.isEmpty() ? ContentType { ThreadableBlobRegistry::blobType(url) } : contentType, keySystem, !!m_remotePlaybackConfiguration))
+    bool needContentType =
+#if ENABLE(MEDIA_SOURCE)
+        !m_mediaSource &&
+#endif
+#if ENABLE(MEDIA_STREAM)
+        !m_mediaStreamSrcObject &&
+#endif
+        !m_remotePlaybackConfiguration;
+
+    if (needContentType && contentType.isEmpty()) {
+        // Attempt to determine the content-type from the URL file extension.
+        contentType = ContentType { emptyString(), url };
+        if (contentType.isEmpty()) {
+            sniffForContentType(url, WTFMove(completionHandler));
+            return;
+        }
+    }
+    if (!completionHandler(contentType))
         mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
+    else
+        mediaPlayerRenderingModeChanged();
+}
 
-    mediaPlayerRenderingModeChanged();
+void HTMLMediaElement::sniffForContentType(const URL& url, Function<bool(const ContentType&)>&& completionHandler)
```

Comment:
I'm curious why this takes a `Function` rather than returning a `NativePromise`?

Reply:
it just made it easier in this context, it also avoids an extra dispatch to stay closer to the existing implementation.

Reply:
It is weird to have an async function that takes in a completion handler not consistently call its completion handler. I think this should take an actual `WTF::CompletionHandler` type is should be called consistently. It's really not a great API otherwise.

Reply:
point noted